### ISSUE TITLE
In 3D, effects not coinciding with camera preview 

### ISSUE THREAD

**Issue by skyjumprr** [smeighan/xLights#1595]
I am using version 2019.32 but this issue is reproducible in earlier versions as well. 
I have two views, Normal and South. My normal view is from the front and I have found not issues with that view as it works like default. My issue is when I try to render an effect on the south camera view I get weird results both with single elements and whole house (or view) effects. 
Here is a link to a zoom video which explains this better.  
[3DTest issue.zip](https://github.com/smeighan/xLights/files/3348249/3DTest.issue.zip)

 Also here is a link to the packaged sequence so you can perhaps reproduce my issue. Or tell me what I'm doing wrong.
[3d test.zip](https://github.com/smeighan/xLights/files/3348234/3d.test.zip)

As this is my first submission for an issue I hope I'm doing it right... If you have any questions or need any more info, please dont hesitate to contact me. at email skyjumpr@toast.net

Thank you for all your work on this awesome software!
Laron

**Comment by keithsw1111** [comment]
Yeah ... I was demoing this the other weekend and it didnt look right then ... I hate debugging this code.

**Comment by AzGilrock** [comment]
I can try to look at this sometime.  To help do some legwork skyjumpr it would help if you could test in version 2019.01 when 3D first came out and see if it's doing the same thing.  If there is a change then you use the divide and conquer method to figure out which release it changed in.  Like use 2019.15 next and then goto 2019.7 or 2019.22 depending on whether the issue is present.  You can install the xLights releases side-by-side.  I usually just type in the version number on the end of the folder name during installation.

**Comment by skyjumprr** [comment]
First I want to thank you both for responding to my issue so quickly!  
Gil, I found this to be an issue from the beginning as per the camera view not syncing with the effect on the  "south house" camera view. However, it was different. The differences changed in version 15 to what they currently are in 32.
 Also found that in versions 1-14 the matrix (or perhaps any single element) the "default" render style was the same as my "per preview, south house" which seems correct. But 15-32, "default" is the only render style that gives expected results from my "south house" camera view.
The pinwheel reveals some interesting results on the matrix as well.  For example, the pinwheel looks closer to correct using the normal(front) camera view instead of the "south house" which is directly in front of the matrix. 
What more can I do to help diagnose this issue?

Thank you
Laron Leake

**Comment by keithsw1111** [comment]
From the readme:

2019.15 March 2, 2019
   -- bug (keith)   Fix rendering per preview from a 3D camera ... it just wasn't right This may change how some rendering looks. I am happy to look into any issues but I will only change it again if it is clearly wrong.

**Comment by AzGilrock** [comment]
Yeah it looks like that change was wrong.  I did a fill effect from a Top view and it's filling every tree at once like its using Per Model Per Preview instead of Per Preview.
![image](https://user-images.githubusercontent.com/2024669/60614921-cbae0700-9d82-11e9-9a9e-e2d8b1090c03.png)

In 2019.12 (closest I had to 15) it shows the Fill effect going from white to red as it swept across the view.
![image](https://user-images.githubusercontent.com/2024669/60615104-3c552380-9d83-11e9-9054-5b88359e8490.png)

**Comment by AzGilrock** [comment]
So when I test in 2019.1 it appears to work as I expected with just one quirk.  The buffer it creates appears to place all the models at the bottom of the buffer so you end up with a lot of blank space on the top side of the buffer.  Its easiest to see if you use a Fill Effect where you can use a ramp to make it go up and down and you'll see after it reaches the top of the models it takes some time before it comes back down because it's up there in that blank space.

**Comment by keithsw1111** [comment]
When i worked on the .15 change i was more focussed on a side-ish view ... so I wonder if there are different things to consider here.

**Comment by skyjumprr** [comment]
Am I trying to do something that the 3d wasn't actually created to do? Is there a better way than the approach I'm attempting? I am definitely not a "power user", so I'm curious who else is having the same issue or just haven't noticed it. I definitely no nothing about code,  but if there is anything I can do to help, please let me know. On Jul 3, 2019 5:27 PM, Keith Westley <notifications@github.com> wrote:When i worked on the .15 change i was more focussed on a side-ish view ... so I wonder if there are different things to consider here.

**Comment by AzGilrock** [comment]
skyjumprr - Keith and I are just talking to each other here you weren't doing anything wrong.  It should work the way you expected.

Keith - I haven't really looked at what got changed for .15.  Haven't been feeling well past few days but I'll try to get to it soon.  The way I saw it work in the earlier versions looked like it would have been good if we have taken the buffer and reduced it to more of a minimal grid to get rid of the extra space around the top.

**Comment by keithsw1111** [comment]
I dont remember all the details but I changed it because of similar issues where effects using the camera just didnt look right ... and i swear it was better ... but clearly it isnt fixed now. This is all very confusing.

I can easily undo the change ... but then i am concerned to verify that the problem i was seeing doesn't come back.

Maybe that problem was with per model per preview ... but i dont think it was. If you dont get to it i will likely do so over the weekend.

**Comment by AzGilrock** [comment]
What's interesting is my SmartGit has not been updated since that checkin.  Probably was the thing that caused me to take a break.
![image](https://user-images.githubusercontent.com/2024669/60685911-a7255e00-9e5a-11e9-90c9-42776b103d01.png)

**Comment by AzGilrock** [comment]
So there were 2 checkins related to this I believe.  The first one deleted the line where it skipped model groups but the reason it was skipping them is because they are already processed by that point.  The second checkin on March 1st deleted that big chunk of code that processes the model groups.  That section was designed so that nested model groups would work.  When it hits a model group it recursively calls itself for each model inside the model group.  Without that section model groups more than 1 deep won't work and because it's already recursed all the models that's why it skips models groups later because it's already handled every model individually within the group.  I think it needs those sections put back and then analyze a way to trim blank space in the buffer edges.  Without doing that the center point will look off and that might be what made it look wrong in the first place.  I'll try to experiment.

**Comment by AzGilrock** [comment]
Not sure why it hasn't been a problem for anyone else but I pulled the code and FPP.cpp wouldn't compile unless I added #include <cctype> because of the std::isspace function.

**Comment by AzGilrock** [comment]
I notice models don't paste in place when you copy/paste now.  I'm not a fan of the offset.  I used to like to paste in place and then drag them along one axis and get perfect alignment.  Now they paste and move down in Y below ground level and shift to the right.

**Comment by computergeek1507** [comment]
Fpp.cpp compiled in vs2019 for me but I have to change std::isspace to issppace for vs2017 to compile.

**Comment by AzGilrock** [comment]
Yeah I'm using vs2017.  If you add the include it fixes it also.  BTW I think I have this issue fixed.

**Comment by AzGilrock** [comment]
So I added back in some of the original code and I found the issue with it was that the nodes were being added to the Nodes object twice.  Once for the model group and again for each model.  By skipping the add if its a 3D render view buffer it keeps it from messing up the min/max values.  Also since its recursive it needed to not do the scale factor adjustment to the node until it was the final pass for the model group.
